wingtips-servlet-api/src/main/java/com/nike/wingtips/servlet/WingtipsRequestSpanCompletionAsyncListener.java

```diff
@@ -23,59 +32,130 @@
 class WingtipsRequestSpanCompletionAsyncListener implements AsyncListener {
 
     protected final TracingState originalRequestTracingState;
-    protected boolean alreadyCompleted = false;
+    // Used to prevent two threads from trying to close the span at the same time 
+    protected final AtomicBoolean alreadyCompleted = new AtomicBoolean(false);
+    protected final HttpTagAndSpanNamingStrategy<HttpServletRequest, HttpServletResponse> tagAndNamingStrategy;
+    protected final HttpTagAndSpanNamingAdapter<HttpServletRequest, HttpServletResponse> tagAndNamingAdapter;
 
-    WingtipsRequestSpanCompletionAsyncListener(TracingState originalRequestTracingState) {
+    WingtipsRequestSpanCompletionAsyncListener(
+        TracingState originalRequestTracingState,
+        HttpTagAndSpanNamingStrategy<HttpServletRequest, HttpServletResponse> tagAndNamingStrategy,
+        HttpTagAndSpanNamingAdapter<HttpServletRequest, HttpServletResponse> tagAndNamingAdapter
+    ) {
         this.originalRequestTracingState = originalRequestTracingState;
+
+        // We should never get a null tag strategy or tag adapter in reality, but just in case we do, default to
+        //      the no-op impls.
+        if (tagAndNamingStrategy == null) {
+            tagAndNamingStrategy = NoOpHttpTagStrategy.getDefaultInstance();
+        }
+
+        if (tagAndNamingAdapter == null) {
+            tagAndNamingAdapter = NoOpHttpTagAdapter.getDefaultInstance();
+        }
+
+        this.tagAndNamingStrategy = tagAndNamingStrategy;
+        this.tagAndNamingAdapter = tagAndNamingAdapter;
     }
 
     @Override
-    public void onComplete(AsyncEvent event) throws IOException {
-        completeRequestSpan();
+    public void onComplete(AsyncEvent event) {
+        completeRequestSpan(event);
     }
 
     @Override
-    public void onTimeout(AsyncEvent event) throws IOException {
-        completeRequestSpan();
+    public void onTimeout(AsyncEvent event) {
+        /*
+            NOTE: Unfortunately the response in the event won't have had its final HTTP status code set yet, so if we
```

Comment:
@adriancole I think Brave's `TracingAsyncListener` is probably also affected by this - it causes the span to have incorrect HTTP response status codes in the case of async timeout or error. Solution is to have `onTimeout(...)` and `onError(...)` do nothing, as according to the Servlet spec `onComplete(...)` should be called afterward, with proper status code set.

See the inline code comment here for full details.

(Yay servlet! 🙄  )